qiskit/circuit/singleton_gate.py

```diff
@@ -0,0 +1,177 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Singleton gate classes.
+"""
+import copy
+
+from qiskit.circuit.gate import Gate
+from qiskit.circuit.classicalregister import ClassicalRegister, Clbit
+from qiskit.circuit.exceptions import CircuitError
+
+
+class SingletonGate(Gate):
+    """A base class to use for Gate objects that by default are singleton instances
+
+    This class should be used for gate classes that have fixed definitions and
+    do not contain any unique state. The canonical example of something like
+    this is :class:`~.HGate` which has an immutable definition and any
+    instance of :class:`~.HGate` is the same. Using singleton gates
+    as a base class for these types of gate classes provides a large
+    advantage in the memory footprint of multiple gates.
+
+    The exception to be aware of with this class though are the :class:`~.Gate`
+    attributes :attr:`.label`, :attr:`.condition`, :attr:`.duration`, and
+    :attr:`.unit` which can be set differently for specific instances of gates.
+    For :class:`~.SingletonGate` usage to be sound setting these attributes
+    is not available and they can only be set at creation time. If any of these
+    attributes are used instead of using a single shared global instance of
+    the same gate a new separate instance will be created.
+    """
```

Comment:
```suggestion
    attributes are used, then instead of using a single shared global instance
    of the same gate a new separate instance will be created.
    """
```

Reply:
Fixed in: https://github.com/Qiskit/qiskit/pull/10314/commits/9c9bc0999f892bcf558fff52629bc89471b3f414